src/Compilers/CSharp/Portable/Symbols/Source/SourceModuleSymbol.cs

```diff
@@ -529,6 +533,86 @@ protected override void DecodeWellKnownAttributeImpl(ref DecodeWellKnownAttribut
             }
         }
 
+#nullable enable
+
+        internal bool RequiresRefSafetyRulesAttribute()
+        {
+            if (_lazyRequiresRefSafetyRulesAttribute == ThreeState.Unknown)
+            {
+                bool value = UseUpdatedEscapeRules && namespaceIncludesRefs(GlobalNamespace);
+                _lazyRequiresRefSafetyRulesAttribute = value.ToThreeState();
+            }
+            return _lazyRequiresRefSafetyRulesAttribute.Value();
+
+            static bool namespaceIncludesRefs(NamespaceSymbol ns)
+            {
+                foreach (var member in ns.GetMembersUnordered())
+                {
+                    switch (member.Kind)
+                    {
+                        case SymbolKind.Namespace:
+                            if (namespaceIncludesRefs((NamespaceSymbol)member))
+                            {
+                                return true;
+                            }
+                            break;
+                        case SymbolKind.NamedType:
+                            if (typeIncludesRefs((NamedTypeSymbol)member))
+                            {
+                                return true;
+                            }
+                            break;
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(member.Kind);
+                    }
+                }
+                return false;
+            }
+
+            static bool typeIncludesRefs(NamedTypeSymbol type)
+            {
+                if (type.IsRefLikeType)
+                {
+                    return true;
+                }
+                foreach (var member in type.GetMembersUnordered())
+                {
+                    switch (member.Kind)
+                    {
+                        case SymbolKind.NamedType:
+                            if (typeIncludesRefs((NamedTypeSymbol)member))
+                            {
+                                return true;
+                            }
+                            break;
+                        case SymbolKind.Method:
+                            if (methodIncludesRefs((MethodSymbol)member))
+                            {
+                                return true;
+                            }
+                            break;
+                    }
+                }
+                return false;
+            }
+
+            static bool methodIncludesRefs(MethodSymbol method)
+            {
+                if (method.RefKind != RefKind.None ||
+                    method.ReturnType.IsRefLikeType)
+                {
+                    return true;
+                }
+                return method.Parameters.Any(p => parameterIncludesRef(p));
+            }
+
+            static bool parameterIncludesRef(ParameterSymbol parameter)
+            {
+                return parameter.RefKind != RefKind.None ||
```

Comment:
We want to return true for `ref string` parameters?

Reply:
We could narrow this further, to reduce the number of cases where we emit a `[module: RefSafetyRules]`. We could also go the other way and emit `[module: RefSafetyRules]` for every assembly compiled with .NET 7 or `-langversion:11`. The current approach was a middle ground but I'm open to suggestions on alternatives.